### PR TITLE
[13] Create internal logging tool

### PR DESCRIPTION
**Pull request by kanat**
### 🎯 Goal

Ticket: https://github.com/GetStream/android-internal-board/issues/13

Introduce `StreamLog` - a new parameterized file logging.

### 🛠 Implementation details

The logging is getting enabled by default in debug builds only, cause `stream-logger-android` module is getting linked as a dependency using `debugApi`.

`stream-logger-android` module contains `StreamLogFileProvider` class which is getting invoked right before `Application.onCreate()`. This class initializes the `StreamLog` with android & file logging.

### 🧪 Testing

- Cut new debug build of any sample app.
- Launch app 
- send in terminal `adb shell am startservice -a io.getstream.logging.android.CLEAR`
- you should see the toast message `Logs cleared`
- open any channel and send a few messages
- send in terminal `adb shell am startservice -a io.getstream.logging.android.SHARE`
- you should see a file-sharing dialog chooser.
- share log file
- check it's not empty & contains log content.

##### Option 1:
```kotlin
private const val TAG = "SomeViewModel"
// ...
StreamLog.d(TAG) { "[<method_name>] arg1: $arg1, arg2: $arg2, arg3: $arg3" }
```
##### Option 2:
```kotlin
val logger = StreamLog.getLogger("SomeViewModel")
// ...
logger.d { "[<method_name>] arg1: $arg1, arg2: $arg2, arg3: $arg3" }
```

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [ ] ~~Changelog is updated with client-facing changes~~
- [ ] ~~New code is covered by unit tests~~
- [ ] ~~Comparison screenshots added for visual changes~~
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_

### CODE REVIEW

File: libraries/logging/stream-logger-android/build.gradle
Comment (adasiewiczr) [inline libraries/logging/stream-logger-android/build.gradle:14]: Do we want to publish those libraries?

File: libraries/logging/stream-logger-android/detekt-baseline.xml
Comment (adasiewiczr) [inline libraries/logging/stream-logger-android/detekt-baseline.xml:6]: Can we fix those issues before merging?

File: libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/StreamLogFileManager.kt
Comment (adasiewiczr) [inline libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/StreamLogFileManager.kt:22]: Can we add KDocs to this object, methods, etc?

File: libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/impl/LifecycleAwareLogFileManager.kt
Comment (adasiewiczr) [inline libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/impl/LifecycleAwareLogFileManager.kt:29]: Let's add KDoc here

File: libraries/logging/stream-logger/src/main/java/io/getstream/logging/CompositeStreamLogger.kt
Comment (adasiewiczr) [inline libraries/logging/stream-logger/src/main/java/io/getstream/logging/CompositeStreamLogger.kt:19]: Let's add KDoc

File: libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLog.kt
Comment (adasiewiczr) [inline libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLog.kt:25]: Missing KDoc

File: stream-chat-android-core/build.gradle
Comment (adasiewiczr) [inline stream-chat-android-core/build.gradle:34]: Shouldn't we include it only as `debugApi`?

File: stream-chat-android-client/build.gradle
Comment (adasiewiczr) [inline stream-chat-android-client/build.gradle:61]: I think we don't need to include this here as it was already added in `core` module

**Review by adasiewiczr — COMMENTED** [review]
Great job overall! I left some small comments and questions.
Would love to get second review from @JcMinarro

**Review by adasiewiczr — COMMENTED** [review]
One more thing - we should update `.github/CODEOWNERS` file

File: stream-chat-android-core/build.gradle
Comment (kanat) [reply stream-chat-android-core/build.gradle:34]: `core` contains a dependency on a `stream-logger` which is just an entry point and does not contain any logging behind.

`StreamLog` is disabled by default

File: stream-chat-android-client/build.gradle
Comment (kanat) [reply stream-chat-android-client/build.gradle:61]: This one brings an additional dependency `stream-logger-android` which contains `Logcat` wrapper, `File` wrapper and also enables the logging

File: libraries/logging/stream-logger-android/build.gradle
Comment (JcMinarro) [reply libraries/logging/stream-logger-android/build.gradle:14]: I think we will need it as well we need it on the `core` module.
It won't be a "direct" dependency our customers will use, but a transitive dependency that could be needed

File: libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/impl/ActivityLifecycleCallbacks.kt
Comment (JcMinarro) [inline libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/impl/ActivityLifecycleCallbacks.kt:23]: Do we really need this abstract class? :thinking:

File: libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/impl/LifecycleAwareLogFileManager.kt
Comment (JcMinarro) [inline libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/impl/LifecycleAwareLogFileManager.kt:33]: Instead of been using thread, couldn't we use coroutines?

File: libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt
Comment (JcMinarro) [inline libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt:66]: Perform logic, and on this case background tasks, is not a good idea.
As an example, if we would need to mock/spy this class in our tests, we won't be able to avoid this logic, that maybe is not needed for our tests.
Could you considere to extract it outside of the class??
One way could be to have a private constructor that receive all needed arguments, and having a companion method to build this an instance of this class

File: libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt
Comment (JcMinarro) [inline libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt:49]: Could we inject this collaborator by constructor?
On the other hand, could we move from "Threads" to "Coroutines"?

File: libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt
Comment (JcMinarro) [inline libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt:112]: Could you add the return type of this function?

File: libraries/logging/stream-logger/detekt-baseline.xml
Comment (JcMinarro) [inline libraries/logging/stream-logger/detekt-baseline.xml:8]: Can we fix those issues before merging?

File: libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLog.kt
Comment (JcMinarro) [inline libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLog.kt:35]: It doesn't seem an `init()` but a `set()` method.
An `init()` method is expected to be called at the beginning and only once, and if by some reason it is called twice, notify about it.
WDYT about to rename it?

Comment (JcMinarro) [reply libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLog.kt:35]: On the other hand, we we need to update both params in one shot?
Maybe we can only use `var` fields and allow to update it from outside

File: libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLogger.kt
Comment (JcMinarro) [inline libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLogger.kt:39]: What about to have an enum class for those type?

File: libraries/logging/stream-logger/src/main/java/io/getstream/logging/helper/stringify.kt
Comment (JcMinarro) [inline libraries/logging/stream-logger/src/main/java/io/getstream/logging/helper/stringify.kt:37]: With an enum, this else wouldn't be needed

File: libraries/logging/stream-logger/src/main/java/io/getstream/logging/kotlin/KotlinStreamLogger.kt
Comment (JcMinarro) [inline libraries/logging/stream-logger/src/main/java/io/getstream/logging/kotlin/KotlinStreamLogger.kt:41]: This logic seems to be duplicated across the PR, maybe we could extract it

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerImpl.kt
Comment (JcMinarro) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerImpl.kt:550]:
```suggestion
            logger.d { "[handleEventsInternal] chatEvent: $chatEvent" }
```

**Review by JcMinarro — CHANGES_REQUESTED** [review]

File: libraries/logging/stream-logger-android/detekt-baseline.xml
Comment (kanat) [reply libraries/logging/stream-logger-android/detekt-baseline.xml:6]: Those 2 are produced because of the following things:

1. **EmptyFunctionBlock**
```kotlin
internal abstract class ActivityLifecycleCallbacks : Application.ActivityLifecycleCallbacks {

    override fun onActivityCreated(activity: Activity, bunlde: Bundle?) {}

    override fun onActivityStarted(activity: Activity) {}

    override fun onActivityResumed(activity: Activity) {}

    override fun onActivityPaused(activity: Activity) {}

    override fun onActivityStopped(activity: Activity) {}

    override fun onActivitySaveInstanceState(activity: Activity, bunlde: Bundle) {}

    override fun onActivityDestroyed(activity: Activity) {}
}
```

2. **SwallowedException**
```
private fun Context.getVersionName(): String = try {
    packageManager?.getPackageInfo(packageName, 0)?.versionName ?: ""
} catch (e: PackageManager.NameNotFoundException) {
    ""
}
```

File: libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/impl/ActivityLifecycleCallbacks.kt
Comment (kanat) [reply libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/impl/ActivityLifecycleCallbacks.kt:23]: Wanna keep `LifecycleAwareLogFileManager` clean from redundant functions

File: libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/AndroidStreamLogger.kt
Comment (kanat) [reply libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/AndroidStreamLogger.kt:36]: Yup, that's true

Screenshot 2022-04-22 at 13 49 49

File: libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLogger.kt
Comment (kanat) [reply libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLogger.kt:39]: I keep as `Int` to make logic like
```kotlin
val validator = IsLoggableValidator { priority: Int, tag: String ->  
    priority > StreamLogger.INFO && tag in validTags
}
```

File: libraries/logging/stream-logger/src/main/java/io/getstream/logging/helper/stringify.kt
Comment (kanat) [reply libraries/logging/stream-logger/src/main/java/io/getstream/logging/helper/stringify.kt:37]: Answered above why I keep `Int`

File: libraries/logging/stream-logger/src/main/java/io/getstream/logging/kotlin/KotlinStreamLogger.kt
Comment (kanat) [reply libraries/logging/stream-logger/src/main/java/io/getstream/logging/kotlin/KotlinStreamLogger.kt:41]: the `composed` is different in diff Loggers

**AndroidLogger**: 
```kotlin
val composed = "($thread) $message"
```

**StreamLogger**: 
```kotlin
val composed = "$now ($thread) [${priority.stringify()}/$tag]: $message"
```

File: libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt
Comment (kanat) [reply libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt:66]: There is a specified implementation for Tests named `KotlinStreamLogger`

File: libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/impl/LifecycleAwareLogFileManager.kt
Comment (kanat) [reply libraries/logging/stream-logger-android/src/main/java/io/getstream/logging/android/file/impl/LifecycleAwareLogFileManager.kt:37]: It's a fully android specific class and the idea is to keep this class off 3rd party libraries as coroutines or rxjava and use android SDK only.

File: libraries/logging/stream-logger-android/detekt-baseline.xml
Comment (JcMinarro) [reply libraries/logging/stream-logger-android/detekt-baseline.xml:6]: If we really need those empty methods (Maybe we don't really need this empty class and move the implementation to the class that implement this abstract class), we should add a comment like `// no-op`, with it we are sure it was thought to not do anything and it is not because we forgot to implement it

For the second one, as far as we are not using the exception, it can be called `_` (default Kotlin's name for ignored arguments) or contain `ignored` as a part of the name

File: libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt
Comment (kanat) [reply libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt:49]: I wanna keep this module clean from any 3rd party libs

Comment (kanat) [reply libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt:49]: yes we can move it constructor

File: libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt
Comment (JcMinarro) [reply libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt:66]: It is fine that we have another implementation that could be used on tests, but, what about classes that explicitly ask for a `FileStreamLoger`? It can't be mocked without running this logic.
Keeping `init()` block outside of our classes are good practiced

File: libraries/logging/stream-logger/detekt-baseline.xml
Comment (kanat) [reply libraries/logging/stream-logger/detekt-baseline.xml]: `EmptyFunctionBlock` comes from
```kotlin
public object SilentStreamLogger : StreamLogger {

    override fun log(priority: Priority, tag: String, message: String, throwable: Throwable?) {}
}
```

Comment (kanat) [reply libraries/logging/stream-logger/detekt-baseline.xml]: Others are fixed

File: libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt
Comment (JcMinarro) [reply libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt:49]: We can use a default params with the `Executors.newSingleThreadExecutor()` value, but allowing to inject a different one on the case the one that will use this class needs it

File: libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLogger.kt
Comment (JcMinarro) [reply libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLogger.kt:39]: You can have almost the same logic with enum:
```
enum class Priority(val level: Int) {
    VERBOSE(2),
    DEBUG(3),
    INFO(4),
    WARN(5),
    ERROR(6),
    ASSERT(7),
}

val validator = IsLoggableValidator { priority: Priority, tag: String ->
    priority.level > Priority.INFO.level
}
```
We can create a function into our enum class to return if the level is bigger

Comment (JcMinarro) [reply libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLogger.kt:39]: Using a "real type", other devs can know which values they can use without digging into the class to see which values are valid and which one not, allowing us to refactor it in the future without breaking our customers logic

File: libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt
Comment (kanat) [reply libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt:66]: `FileStreamLoger` is used at `StreamLogFileProvider` only. That ContentProvider will be included in Debug builds only. There is no need to mock `FileStreamLoger`.

File: libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt
Comment (kanat) [reply libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt:49]: But I don't wanna allow passing any diff Executor by design to restrict it to be always a SingleThread one.

File: libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLogger.kt
Comment (kanat) [reply libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLogger.kt:39]: What is a big advantage of turing it into `enum`?
I won't use enum as it is cause I still will be converting names to:
```kotlin
 VERBOSE -> "V"
```

Comment (JcMinarro) [reply libraries/logging/stream-logger/src/main/java/io/getstream/logging/StreamLogger.kt:39]: What is the advantage of using `typealias Priority = Int`? 
It does nothing because we don't limit what you can use to define the log level.
Keeping a good type structure on our code helps us to reduce bugs in the future.
If you use Int, and for some reason, we need to introduce an intermediate value, it will break all our customer's apps.
Instead of that, if we have a clear data structure (In this case, an enum type), it won't break our customer's apps because they would be using "an Int value" that doesn't make sense

File: libraries/logging/stream-logger-android/detekt-baseline.xml
Comment (JcMinarro) [reply libraries/logging/stream-logger-android/detekt-baseline.xml:6]: It is ok you create the `ActivityLifecycleCallbacks` class, but it needs to follow our Detek Rules

**Review by JcMinarro — APPROVED** [review]

**Review by adasiewiczr — APPROVED** [review]

**Pull request merged**
